### ISSUE TITLE
New check: implicit global variable declaration

### ISSUE THREAD

**Issue by xzfc** [koalaman/shellcheck#1016]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```bash
func1() {
	local var3
	declare -g var4

	var1=x # Bad: implicit global
	for var2 in {1..10};do : $var2; done # Bad: implicit global
	var3=x # Good: explicit local
	var4=x # Good: explicit global

	func2 # see below

	: $var1 $var3 $var4 # suppress SC2034
}
```

#### Here's what I wanted or expected to see:
Warning about implicit global declaration of `$var1` and `$var2`.

---

However, following are valid cases (assuming that `func2` is only called from `func1`).
I am not sure how ShellCheck should treat this.

```bash
func2() {
	var3=y # modify local variable of `func1`
	var4=y # modify already defined global variable
}
```

---

Related: `zsh` have `setopt WARN_CREATE_GLOBAL` to print waring message when implicit global variable declaration occurs.

In the said example, `zsh` will print followng warning messages:

```
func1:4: scalar parameter var1 created globally in function func1
func1:5: scalar parameter var2 created globally in function func1
```

**Comment by xzfc** [comment]
Oops, just found related issue: #116.

**Comment by Blzut3** [comment]
I also would love to see this a warning for using variables without declaring them.  I've taken to writing a lot of code in the form of:
```bash
func() {
    # Name arguments
    declare ArgName=$1
    shift
    
    # Acquire resource
    mkdir "$ArgName"

    # Call out to another function
    do_stuff_with_resource
    declare Ret=$?

    # Release resource
    rm -rf "$ArgName"

    return "$Ret"
}
```
This form gives me semantics where if an error occurs the script will clean up whatever temporary resources it constructed as the call stack unwinds.  The problem I occasionally run into is that argument names get reused and while I've made it a habit to declare all of my variables before use other people making fixes don't necessarily have this habit.  Plus there are cases where bash doesn't allow a declaration to be at the location of first use including for loops, read, and checking the error code while assigning.  On that note, reusing for loop iterator variable names is another common case where deep call stacks have a chance of causing non-obvious behavior.

I believe the false positive concerns in the old ticket are a non-issue since as xzfc stated there one can use `declare -g` where the global behavior is desired and system environment variables could probably be satisfied by requiring the user to declare them as `export` before use.  I have yet to run into a case where declare before use is not possible and I do believe that my code quality has been significantly improved by it.
